In [57]:
# Packages for data cleaning, plotting, and manipulation
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

# scikit-learn libraries/functions/classes
from sklearn.compose import ColumnTransformer
from sklearn.dummy import DummyClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.impute import SimpleImputer
from sklearn.linear_model import LogisticRegression, RidgeCV
from sklearn.metrics import accuracy_score, precision_score, recall_score
from sklearn.model_selection import train_test_split, GridSearchCV, cross_val_score
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.tree import DecisionTreeClassifier

In [35]:
# Importing training data
# Only using `status_group` column from label set, to
# avoid duplicating `id` column
df_train_val = pd.read_csv('data/training_set_values.csv')
df_train_label = pd.read_csv('data/training_set_labels.csv',
                             usecols = ['status_group'])


df_test = pd.read_csv('data/test_set_values.csv')

In [60]:
# Concatenating separate .csv files
df_train = pd.concat(objs = [df_train_val, df_train_label],
                     axis = 1)

df_train.head()

id  amount_tsh date_recorded        funder  gps_height     installer  \
0  69572      6000.0    2011-03-14         Roman        1390         Roman   
1   8776         0.0    2013-03-06       Grumeti        1399       GRUMETI   
2  34310        25.0    2013-02-25  Lottery Club         686  World vision   
3  67743         0.0    2013-01-28        Unicef         263        UNICEF   
4  19728         0.0    2011-07-13   Action In A           0       Artisan   

   longitude   latitude              wpt_name  num_private  ... water_quality  \
0  34.938093  -9.856322                  none            0  ...          soft   
1  34.698766  -2.147466              Zahanati            0  ...          soft   
2  37.460664  -3.821329           Kwa Mahundi            0  ...          soft   
3  38.486161 -11.155298  Zahanati Ya Nanyumbu            0  ...          soft   
4  31.130847  -1.825359               Shuleni            0  ...          soft   

  quality_group      quantity  quantity_group                source  \
0          good        enough          enough                spring   
1          good  insufficient    insufficient  rainwater harvesting   
2          good        enough          enough                   dam   
3          good           dry             dry           machine dbh   
4          good      seasonal        seasonal  rainwater harvesting   

            source_type source_class              waterpoint_type  \
0                spring  groundwater           communal standpipe   
1  rainwater harvesting      surface           communal standpipe   
2                   dam      surface  communal standpipe multiple   
3              borehole  groundwater  communal standpipe multiple   
4  rainwater harvesting      surface           communal standpipe   

  waterpoint_type_group    status_group  
0    communal standpipe      functional  
1    communal standpipe      functional  
2    communal standpipe      functional  
3    communal standpipe  non functional  
4    communal standpipe      functional  

[5 rows x 41 columns]

In [75]:
# Checking for null values in features columns

df_train.isna().sum()

id                           0
amount_tsh                   0
date_recorded                0
funder                    3635
gps_height                   0
installer                 3655
longitude                    0
latitude                     0
wpt_name                     0
num_private                  0
basin                        0
subvillage                 371
region                       0
region_code                  0
district_code                0
lga                          0
ward                         0
population                   0
public_meeting            3334
recorded_by                  0
scheme_management         3877
scheme_name              28166
permit                    3056
construction_year            0
extraction_type              0
extraction_type_group        0
extraction_type_class        0
management                   0
management_group             0
payment                      0
payment_type                 0
water_quality                0
quality_

Columns/features with large numbers of null values, from least to greatest quantity:

- `subvillage` (371)
- `permit` (3056)
- `public_meeting` (3334)
- `funder` (3635)
- `installer` (3655)
- `scheme_management` (3877)
- `scheme_name` (28166) **(!)**

In [58]:
# Checking datatypes of columns

df_train.dtypes;

In [63]:
# Splitting DataFrame into features/values DataFrame
# (i.e. `X`) and labels series (`y`)

X = df_train.drop('status_group', axis = 1)
y = df_train['status_group']

In [72]:
X.loc[X['funder'].isna()]

id  amount_tsh date_recorded funder  gps_height installer  \
34     41583         0.0    2011-02-23    NaN         -41       NaN   
43     19282         0.0    2013-01-15    NaN        1642       NaN   
47     13620         0.0    2011-07-27    NaN           0       NaN   
65     51072         0.0    2013-02-09    NaN        1415       NaN   
71     17386         0.0    2011-03-31    NaN           0       NaN   
...      ...         ...           ...    ...         ...       ...   
59357  46563         0.0    2013-02-19    NaN        1635       NaN   
59366  55232         0.0    2013-02-02    NaN        1541       NaN   
59370  14796       200.0    2013-01-29    NaN        1154       NaN   
59376  34716         0.0    2013-02-03    NaN        1581       NaN   
59397  37057         0.0    2011-04-11    NaN           0       NaN   

       longitude  latitude             wpt_name  num_private  ...  \
34     39.812912 -7.889986  Msikitini Wa Ijumaa            0  ...   
43     34.967789 -4.628921         Mvae Primary            0  ...   
47     33.540607 -9.172905           Mahakamani            0  ...   
65     34.621598 -5.173136               Nyambi            0  ...   
71     34.462228 -8.575780    Kwa Manyusi Mlilo            0  ...   
...          ...       ...                  ...          ...  ...   
59357  34.971841 -5.098362              Shabani            0  ...   
59366  34.765729 -5.027725              Joshoni            0  ...   
59370  30.058731 -4.902633       Village Office            0  ...   
59376  34.821039 -5.076258              Nasingo            0  ...   
59397  34.017087 -8.750434              Mashine            0  ...   

      payment_type water_quality quality_group      quantity  quantity_group  \
34       never pay         salty         salty        enough          enough   
43         unknown       unknown       unknown           dry             dry   
47       never pay          soft          good        enough          enough   
65         unknown       unknown       unknown           dry             dry   
71       never pay          soft          good  insufficient    insufficient   
...            ...           ...           ...           ...             ...   
59357      unknown       unknown       unknown           dry             dry   
59366    never pay          soft          good  insufficient    insufficient   
59370      monthly       unknown       unknown       unknown         unknown   
59376      unknown       unknown       unknown           dry             dry   
59397      monthly      fluoride      fluoride        enough          enough   

             source   source_type  source_class     waterpoint_type  \
34     shallow well  shallow well   groundwater           hand pump   
43      machine dbh      borehole   groundwater  communal standpipe   
47           spring        spring   groundwater  communal standpipe   
65      machine dbh      borehole   groundwater  communal standpipe   
71            river    river/lake       surface  communal standpipe   
...             ...           ...           ...                 ...   
59357  shallow well  shallow well   groundwater           hand pump   
59366  shallow well  shallow well   groundwater           hand pump   
59370       unknown         other       unknown               other   
59376  shallow well  shallow well   groundwater               other   
59397   machine dbh      borehole   groundwater           hand pump   

      waterpoint_type_group  
34                hand pump  
43       communal standpipe  
47       communal standpipe  
65       communal standpipe  
71       communal standpipe  
...                     ...  
59357             hand pump  
59366             hand pump  
59370                 other  
59376                 other  
59397             hand pump  

[3635 rows x 40 columns]

In [64]:
# Splitting internal training data into separate
# training and test sets for (eventual) internal validation

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.33, random_state = 666)